File: findOriginalArray.java
import java.util.*;
// Approach 1- Sorting + HashMap
class Solution{ // TC O(n log n) SC O(n)
	public int[] findOriginalArray(int [] changed){
		if (changed.length % 2 == 1) return new int[0];
		Arrays.sort(changed);
		HashMap<Integer,Integer> freq=new HashMap<>();
		for (int num: changed){
			freq.put(num,freq.getOrDefault(num,0)+1);
		}

		int [] original = new int[changed.length/2];
		int index=0;
		for (int num: changed){
			if (freq.get(num)>0){
				freq.put(num,freq.get(num)-1);
				int twiceNum=num*2;
				if(freq.containsKey(twiceNum) && freq.get(twiceNum)>0){
					freq.put(twiceNum,freq.get(twiceNum)-1);
					original[index++]=num;
				}
				else{
					return new int[0];
				}
			}
		}
		return original;
	}
}

// Approach 2- Counting Sort 
class Solution1{ // TC O(n+k) SC O(k)
    public int[] findOriginalArray(int[] changed) {
        // It can't be doubled array, if the size is odd
        if (changed.length % 2 == 1) {
            return new int[0];
        }
        
        int maxNum = 0;
        // Find the max element in the array
        for (int num : changed) {
            maxNum = Math.max(maxNum, num);
        }
        
        int[] freq = new int[2 * maxNum + 1];
        // Store the frequency in the map
        for (int num : changed) {
            freq[num]++;
        }
        
        int[] original = new int[changed.length / 2];
        int index = 0;
        for (int num = 0; num <= maxNum; num++) {
            // If element exists
            if (freq[num] > 0) {
                freq[num]--;
                
                int twiceNum = num * 2;
                if (freq[twiceNum] > 0) {
                    // Pair up the elements, decrement the count
                    freq[twiceNum]--;
                    // Add the original number to answer
                    original[index++] = num;
                    num--;
                } else {
                    return new int[0];
                }
            }
        }
        
        return original;
    }
}

class findOriginalArray{
	public static void main(String[] args) {
		int [] changed={1,3,4,2,6,8};
		Solution s= new Solution();
		int [] answer=s.findOriginalArray(changed);
		System.out.println(Arrays.toString(answer));
		Solution1 s1= new Solution1();
		int [] answer1=s1.findOriginalArray(changed);
		System.out.println(Arrays.toString(answer1));
	}
}

File: findOriginalArray.py
# Approach 1- Sorting + HashMap
def findOriginalArray(changed): # TC O(n log n) // SC O(n)
    if len(changed)%2==1:
        return []

    changed.sort()
    hashmap={}
    for i in changed:
        if i in hashmap:
            hashmap[i]+=1
        else:
            hashmap[i]=1

    original=[]
    for n in changed:
        if hashmap[n]>0:
            hashmap[n]-=1
            twiceNum=2*n
            if twiceNum in hashmap and hashmap[twiceNum]>0:
                hashmap[twiceNum]-=1
                original.append(n)
            else:
                return []

    return original

# Approach 2- Counting Sort 
def findOriginalArray1(changed): # TC O(n+k) // SC O(k)
    if len(changed)%2==1:
        return []

    maxNum=max(changed)
    freq=[0]*(2*maxNum+1)
    for num in changed:
        freq[num]+=1
    original=[0]*(len(changed)//2)
    index=0 
    for i in range(maxNum+1):
        if freq[i]>0:
            freq[i]-=1
            twiceNum=2*i 
            if freq[twiceNum]>0:
                freq[twiceNum]-=1
                original[index]=i 
                index+=1
                i-=1
            else:
                return []

    return original

print(findOriginalArray([1,3,4,2,6,8]))
print(findOriginalArray1([1,3,4,2,6,8]))

